Move all documentation in one central point - Prod4pod 1937

irstavr opened the pull request:
## Scope

In this PR, we provide one central point for all the documentation generated in polyPod.
Now all docs are located under `/docs` folder in a structure similar to the project's.
e.g. the docs of `api` can be found under `/docs/platform/feature-api/api/' dir

To do so, we:
Added a new script to `build.js` to generate all the documentation. This new script reads the subdirectories which already have `doc` script and executes the `jsdoc` or `typedoc` to generate our docs.
 
As the documentation till now was partly working only for .js and .ts files, we installed the `better-docs` tool of `jsdoc` in order to support the `jsx` files for the react components. 
Therefore, we needed to set a `jsdoc.conf.json` under every dir that has both .js and .jsx files to support them both. 
And we corrected all the scripts to make use the config file: `"doc": "jsdoc -c jsdoc.conf.json"`

 ### To remark:
As a reminder for any new code that will need documentation, we'll require to copy the `jsdoc.conf.json` to that dir as well. The only change required is to change the 'name' of the documentation page and the _output_ folder:
e.g. for 'poly-look' we have:
``` .json
    "templates": {
      "better-docs": {
        "name": "_poly-look documentation_" // TBC
      }
    },
    "opts": {
      "template": "../../node_modules/better-docs",
      "destination": "../../docs/feature-utils/poly-look",  // TBC
      "recurse": true,
      "verbose": true,
      "readme": "README.md"
    }
```

Also, for the 'prettiness' of the resulted page, we set the Readme as the global html page, the param "readme" on the script reads the path of the file. In case it's missing, the script will fail. (Thus, a copy of a 'Readme' was added to google feature as there was none till now).

<!-- Link here the related JIRA issue in case of a bug / feature -->
### 🏗️ Fixes ([#Prod4Pod-1937](https://jira.polypoly.eu/browse/PROD4POD-1937))

JJ commented on feature-utils/poly-look/package.json:
```diff
@@ -13,7 +13,7 @@
     "build": "rollup -c rollup.config.mjs",
     "build-downstream": "../../build.js --start @polypoly-eu/poly-look",
-    "doc": "jsdoc -R README.md -r src -d ./docs"
+    "doc": "jsdoc -c jsdoc.conf.json"
```
If this is the only change, how come package-lock.json have changed?

irstavr replied:
I had initially installed the better-docs locally but removed it - seems the .lock was affected but didnt change after the remove package. will revert

JJ replied:
But it's used anyway, right? The templates are used?

irstavr replied:
if you mean this change in package.json, yes it's used. the .lock file is reverted indeed

JJ replied:
No, I meant the better-docs package.

JJ commented on feature-utils/poly-look/src/react-components/buttons/polyButton.jsx:
```diff
@@ -1,12 +1,23 @@
 import React from "react";
 import "./polyButton.css";
 
+const types = {
```
This is a big PR already. It would be better if other changes unrelated to generating the documentation were kept off it...

JJ reviewed (changes requested):
Main blocker here is the package-lock.json thing that seems unrelated. The rest would be nice to address, but no big deal.

irstavr commented on .gitignore:
```diff
@@ -27,3 +27,6 @@ android/app/local.properties
 debug.log
+
+# Generated documentation
+docs/
```
well, till now we werent' publishing the docs directly in github - rather than leave it up to the user to check them out by running the `doc` script

JJ replied:
... which would mean we would eventually have two different directories, with slightly different names, as document folders, causing some confusion (or a lot)
I know we were not publishing the docs so far. But the issue here is what these generated documents are for: the use case is the local developer that will run the script themselves, or do we want to publish them on the web for everyone to see and make the repo more presentable? I think it's 2 what we want, and that implies two decisions: use the docs directory (for this and the architecture documents) *and* not ignore it. Since actually changing the repo configuration it's something you and I cannot do, I think that not adding it to the .gitignore file clearly signals the intent, if that's precisely where we are going.

JJ replied:
(That would, BTW, also remove the test fail that's there)

irstavr replied:
I was thinking the same... to add a `docs/documentation` folder where all generated docs are located (and can be ignored instead of uploading the git) and rest can lie in `docs/` as it is now..

JJ replied:
As long as it's not ignored, I'm good with it. If it's ignored, we would have to delete that later on. But if you think the first option (dev-generated docs) is what's going to win in the end, this is a nice tradeoff, because just removing that entry could change to option 2.

irstavr commented on feature-utils/poly-look/jsdoc.conf.json:
```diff
@@ -0,0 +1,25 @@
+      }
+    },
+    "opts": {
+      "template": "../../node_modules/better-docs",
```
our own tool to handle react components? what do you mean?

JJ replied:
A template that's in the repo, and not generated by the install process.

irstavr replied:
well this would mean a rewriting of `better-docs` tool... if I get you correctly. haven't seen anywhere doing it that way tbh

JJ replied:
OK, I don't think we're on the same page here. So let's start over. What exactly does this configuration do?

irstavr replied:
we need the `better-docs` in order to support the `.jsx` documentation... it's not properly handle by `jsdoc` with what we've till now..

JJ replied:
OK, I get now that what handles the .jsx files is this other module. But I circle back to my initial objection: is that now installed in some path-accessible place? Isn't it run via npx so that you can access it in its binary directory? Can you change that to `npx better-docs` so that you don't depend on specific paths?

irstavr replied:
It didn't work first hand - will look at it further at a next PR as it's fine for you this one

irstavr commented:
> Main blocker here is the package-lock.json thing that seems unrelated. The rest would be nice to address, but no big deal.

Corrected the .lock file and moved the fixing of jdocs to https://github.com/polypoly-eu/polyPod/pull/1233

JJ reviewed (approved):
Main hurdle is removed, so looks good to me. You might want to address my concerns or leave them for later, your call.

irstavr commented:
> Main hurdle is removed, so looks good to me. You might want to address my concerns or leave them for later, your call.

All are fixed. Just https://github.com/polypoly-eu/polyPod/pull/1231#discussion_r1005724412 is left for a future review.
Thank you 🌷

L-E commented:
This change introduced a vulnerable dependency in `pug` (via `better-docs`).
